### PR TITLE
[IMP] Update wrong cache invalidation on ir.property

### PR DESCRIPTION
**Pull request by jbjourget-b**
Directly linked to [#79459](https://github.com/odoo/odoo/issues/79459#issuecomment-1744913884) 

Fixes a bug which causes that: self.flush() and self.clear_caches() were called on every property write/update and also delete.
The same bug also exists in Odoo 14/15/17

Current behavior before PR:
Running a modification or deleting a ir.property is invalidating cache at all time.

Desired behavior after PR is merged:
Only invalidating cache when the ir.property has no res_id or will have no res_id.

This is a real improvement in Odoo performance.

### CODE REVIEW

File: odoo/addons/base/models/ir_property.py [hunk 1]
```diff
@@ -110,12 +110,11 @@ def write(self, values):
         # if any of the records we're writing on has a res_id=False *or*
         # we're writing a res_id=False on any record
         default_set = False
+        values = self._update_values(values)
```
Comment (ryv-odoo) [inline odoo/addons/base/models/ir_property.py:113]: Is it useful ? `_update_values` doesn't change `res_id`. Also, it looks like you call `_update_values` twice now.

**Review by ryv-odoo — COMMENTED** [review]
Thank you @jbjourget-b ,
Indeed, it looks like a big source of cache invalidation.
@Xavier-Do cleaned up the `write` part for V17 with https://github.com/odoo/odoo/pull/119813/files#diff-0b4db946b041bf0f6d6ce027dc7ea067ce4a610deef22fcb90a190aa19402883R117-R126 (the unlink part is missing).

You should target the V15, right?

**Comment by ryv-odoo** [comment]
Also, could you add [CLA](https://github.com/odoo/odoo/blob/master/doc/cla/sign-cla.md) commit and follow the guideline about the [commit message](https://www.odoo.com/documentation/17.0/contributing/development/git_guidelines.html#commit-message-structure). Thank you

File: odoo/addons/base/models/ir_property.py (hunk 1 above)
Comment (jbjourget-b) [reply odoo/addons/base/models/ir_property.py:113]: Indeed the `_update_values` can be done only once before and not in the write

**Comment by jbjourget-b** [comment]
@ryv-odoo 
In V17 the fix also is running in more cases than intended:

`default_set = (
            # turning a record value into a fallback value
            values.get('res_id') is False and any(record.res_id for record in self)
        ) or any(
            # changing a fallback value
            not record.res_id and any(record[fname] != self._fields[fname].convert_to_record(value, self) for fname, value in values.items())
            for record in self
        )`

`values.get('res_id') is False` is has no default value other than False. So if res_id is not in values it will also invalidate the cache.

The fix can be used in V14-15-16

**Comment by Xavier-Do** [comment]
> In V17 the fix also is running in more cases than intended:
> `values.get('res_id') is False` has no default value other than False. So if res_id is not in values it will also invalidate the cache.

I don't get what you mean, the default value of get on a dict is non, so if the value is not in the dict if will be False

```
In [3]: {}.get('res_id') is False
Out[3]: False

In [4]: {'res_id': 1}.get('res_id') is False
Out[4]: False

In [5]: {'res_id': False}.get('res_id') is False
Out[5]: True
```

Or did I miss what you mean?

**Comment by jbjourget-b** [comment]
My bad for not remembering `get()` was returning False and not None. I'll adapt my code too

**Review by ryv-odoo — APPROVED** [review]
Thank you @jbjourget-b ,

@robodoo r+

(Note: it was done a little bit differently [here](https://github.com/odoo/odoo/pull/119813/files#diff-0b4db946b041bf0f6d6ce027dc7ea067ce4a610deef22fcb90a190aa19402883R117-R126), where it also avoids clearing the cache when the same default value is written. But this should almost never happen anyway, then lets keep it that way.)

**Comment by ryv-odoo** [comment]
Hello @jbjourget-b , I appear that our enterprise repository need a small modification (add a flush_model before a SQL query) to work properly. Then I cherry-pick your commit (and modify it a little) and create brand new PR in both repositories => https://github.com/odoo/odoo/pull/145171

**Pull request closed without merging**